### PR TITLE
Pre compute the next card's question

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This is a new version of https://github.com/ankidroid/Anki-Android/pull/6026 

I believe the code is quite simpler. Instead of having a list of ID, I have a list of CardCache. That is, essentially, it's an ID, with potentially one more data, the card loaded and even the question preloaded. Thanks to this, I can preload the card content without needing to actually change anything about the content of the queue.

I actually usually preload two cards, because I do not know in advance whether the next card will be a card in learning or not. 

This also means that if for some unknown reason I got the next card wrong, the worst that will occur is that the next card will need to be loaded before being displayed instead of being preloaded.  I don't expect to get the next card wrong, but at this point I'm not as confident as I used to be. Thanks to David for noticing in the last version of the code that I was preloading the next card as if it were a card that we must review now, ignoring that in the future, a card in learning may become due.

The next task would be to figure out how to preload the medias, and maybe even create partially the HTML document to load MathJax in it if required.

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 1]
```diff
@@ -1875,6 +1880,10 @@ public TaskData doInBackgroundCheckCardSelection(TaskData param) {
         return new TaskData(new Object[] { hasUnsuspended, hasUnmarked});
     }
 
+    public void doInBackgroundPreloadNextCard() {
+        getCol().getSched().preloadNextCard();
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1884]: Is there any reason to expect a failure here specifically ?
If there is no col or no sched there is no reason to do anything specifically that is not done everywhere else.

I can't imagine any reason to fear that preloadNextCard would fail more than getCard, and we are not specifically looking for errors elsewhere. We try to precompute the next card by potentially pre-filling the queues and fetching the first card's question. If we can't find any non-empty queue, then we won't do anything. If there was actually a new card to consider and we didn't see it because of counts or something similar, then loading the next card will be potentially slow, but there is nothing I can do to prevent it

**Comment by Arthur-Milchior** [comment]
No need to hold it anymore

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1884]: I think this is an operation where if `getCol.getSched()` fails for example, then we're fine to log & ignore it, as `answerCard` will return the same card, just without caching.

I'm not sure about the actual operation itself and how it handles exceptions.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java [hunk 2]
```diff
@@ -719,7 +746,7 @@ protected void _resetRevQueue() {
 
 
     @Override
-    protected boolean _fillRev(boolean allowSibling) {
+    protected synchronized boolean _fillRev(boolean allowSibling) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:749]: Why has this been added? What has been added that needs to be protected against?

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 3]
```diff
@@ -674,6 +671,60 @@ protected Card _getCard() {
...
+            return new List[]{mLrnQueue};
+        }
+        return new List[]{};
+    }
+
+    /** pre load the potential next card. It may loads many card because, depending on the time taken, the next card may
+     * be a card in review or not. */
+    public void preloadNextCard() {
+        for (List<? extends Card.Cache> caches: _fillNextCard()) {
+            if (caches.isEmpty()) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:720]: I don't get this logic? If cache 1 is empty, why would you not want to reload cache 2?

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 4]
```diff
@@ -708,7 +759,7 @@ private void _resetNew() {
     }
 
     private void _resetNewQueue() {
-        mNewDids = new LinkedList<>(mCol.getDecks().active());
+        mNewDids = new LinkedList(mCol.getDecks().active());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:762]: Why was this changed?

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 5]
```diff
@@ -3063,8 +3128,8 @@ public void setCurrentCard(@NonNull Card card) {
         mCurrentCardParentsDid = currentCardParentsDid;
         _burySiblings(card);
         // if current card is next card or in the queue
-        mRevQueue.remove(card.getId());
-        mNewQueue.remove(card.getId());
+        mRevQueue.remove(mCol.getCardCache(card.getId()));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:3131]: Alternate implementation: Bake the caching into the data structure, and then you don't complicate the API calls here

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1884]: I added the same protection than the one on answer_card

**Comment by Arthur-Milchior** [comment]
"that moving the caching inside the data structure"
Which data structure ? I'm moving it in the data strucutre Card.Cache. But I assume that's not your point.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:749]: In practice, this "synchronized" is as useles as almost all other "synchronized" because we can never have multiple tasks in background simultaneously. And I believe the scheduler is only called in background.

In theory, _fillRev check if the rev queue is empty or not, if it is empty, it fills it, otherwise it states "It's not empty, there is nothing to do here". So you do not want two instances of _fillRev to be called simultaneously, because it would mean we are potentially filling it twice. Since they are really filling the same object, cards would be duplicate in the queue. That would not be a real trouble because they would be discarded due to self-day -spacing, but still, that's better avoided.

I would agree that synchronizing on the queue would be better. That's actually feasible here, because the queue object never change. Do you want me to do it ?

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:720]: Thanks. My bad. In this case, it should have been "continue". I wrote it when I expected to return a single queue.
Corrected

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:762]: No reason I can remember.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 5 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:3131]: Which data structure ? The queue ?  I don't understand what change you want here

**Comment by david-allison-1** [comment]
I would propose a `CardQueue` data structure, which wraps the `LinkedList<CardCache>`, accepts a `CardId`, and internally handles the conversion to a CardCache, this would allow the API of adding a card to the queue to be simplified to the only thing that matters, while keeping the caching operations simple to understand.

It would also allow a `preloadHead` method, which would make the preload operations easier to understand.

This does add another small class inside the scheduler - so I'd be happy with you saying no here.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java (hunk 2 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:749]: I think we should remove it, until we hit the point that we need to think about multithreading. I think we're taking on a performance loss for no material gain, and that's the opposite of what this PR should accomplish.

**Comment by Arthur-Milchior** [comment]
The problem is that the learning queue contains cards and due date, so it should still be treated differently. Not impossible, but not trivial to do either

**Comment by david-allison-1** [comment]
> The problem is that the learning queue contains cards and due date, so it should still be treated differently. Not impossible, but not trivial to do either

Let's say no for now then

**Comment by Arthur-Milchior** [comment]
Done

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java]: Done

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 6]
```diff
@@ -3063,8 +3149,8 @@ public void setCurrentCard(@NonNull Card card) {
...
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:3152]: This should also be implemented in the card queue.

We need to assume that this does not return false. I'm not sure how

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java]: Why do you need to assume it does not return false ?

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java]: Added in the queue class

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 7]
```diff
@@ -2514,7 +2605,7 @@ protected void _burySiblings(Card card) {
                 }
                 // even if burying disabled, we still discard to give
                 // same-day spacing
-                queue_object.remove(cid);
+                queue_object.remove(mCol.getCardCache(cid));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:2608]: I think this is the final one

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:2608]: Done

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 6 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java]: If it returns false, it's very likely a bug in the scheduler's logic - we shouldn't try to remove a card that doesn't exist.

**Comment by Arthur-Milchior** [comment]
How do you track perf ?

I can tell I see an increase for cards with a lot of field in my collection, but that's not really scientific

**Comment by david-allison-1** [comment]
> How do you track perf ?
> 
> I can tell I see an increase for cards with a lot of field in my collection, but that's not really scientific


`sAnalytics.timing()` I'd guess. @mikehardy likely knows the library better.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 6 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java]: I disagree. We are trying to remove the cards from both new and review queue. The current card may well be in the daily lrn queue or in the subday lrn queue. We don't want to remove cards in learning from the queue, so if we don't find the cards in the queue it only means that it is in learning.

It could also means that the queue had been reset in background and that setCurrentCard was executed after the reset and before the queue is filled

**Comment by Arthur-Milchior** [comment]
Sorry, I'm going to ask more precisely:

* since we are not tracking yet, how can we start tracking and hope to see an improvement without having previous numbers
* Since what I care about is how much time it takes for the user to actually see the new question, how do we ensure that the question is actually seen. The time to actually see the question should take into account the time it takes to load media and mathjax; which does not seems to be easily accessible from java

**Comment by mikehardy** [comment]
> > How do you track perf ?
> > I can tell I see an increase for cards with a lot of field in my collection, but that's not really scientific
> 
> `UsageAnalytics: sAnalytics.timing()` I'd guess. @mikehardy likely knows the library better.

I wrapped Analytics up pretty tightly with regards to Java accessibility because I understand this is a very privacy-minded app, so there is no way to reach in directly and do any sort of analytics except for the methods published, and none of them have to do with timing hits (where you could log the time taken for something and thus infer performance of some action in varying conditions)

To track it at all then UsageAnalytics would need an extension where sendAnalyticsTiming was implemented as a new method, and it worked similar to the "Event" hit (where we log screens shown), but let the caller define a name for the timer, and the time taken and any other details I guess?

The items available are here and the API is "fluent" so it would be used like 

```java
sUsageAnalytics.timing()
    .userTimingCategory(catParam)
    .userTimingLabel(labelParam)
    .userTimingVariableName(variableNameParam)
    .userTimingTime(timeTakenParam)
    .sendAsync();
```
https://www.javadoc.io/doc/net.mikehardy/google-analytics-java/latest/com/brsanthu/googleanalytics/request/TimingHit.html

which are really just mapping to these - the Java API ends up being an HTTP POST to this API:

https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#timing

But you are correct, there is no prior measurement to compare to, so we have no idea whether it represents an improvement or not.

That said, the best time to measure performance is when you created the program, the second best time is now. So if this was added and we started measuring some basic things (from app first line of our code executed - attachBaseContext I would think) to full init, and to deck list displaying, and to deck list displaying with numbers, that would be good. Time from card answer or card flip to card displayed again, also good

**Comment by Arthur-Milchior** [comment]
I don't know whether I'm tired or anything, but I really feel I don't have enough context to understand how to implement those measures today.
Actually, even if there were the code, I don't even know where those measure would be sent and how to access it.

Would I be correct if I suggested that measuring time, if we do it, should be done in another PR and that we can already consider this one even if we can't measure how much it will decrease the time users wait to see the deck list

**Comment by mikehardy** [comment]
I'd peel that whole comment off using the github web UI's "reference in a new issue..." option and go from there, to keep this PR focused. A Timing PR is a different beast really.

**Comment by mikehardy** [comment]
PR merged, timer analytics issue created, no action items pending here

**Comment by ekumlin** [comment]
It looks like this was merged in a month and a half ago and has the 2.13 tag &mdash; yet, on AnkiDroid 2.13.3 (Pixel 3, Android 11), revealing answers or moving to the next card still has a delay (during which no text is shown) of about 1 second when loading the font. Should I open a new bug for this?

**Comment by mikehardy** [comment]
font loading is the responsibility of the webview from the system and is not in our control, if your font loads take a second, there is nothing you can do except use a lighter font - this is unfortunately an often-discussed problem, but there is nothing we can do here. We do not pre-render pages (that would load the font) we pre-compute contents (just getting them ready for render)

**Comment by ekumlin** [comment]
Got it. Thanks Mike.

**Comment by Arthur-Milchior** [comment]
I strongly disagree @mikehardy 
My goal is eventually to preload data in html. Probably by doing a new webviewer that is not displayed immediately. Sadly, given the number of values that are directly in the reviewer, it's not really easy to do right now. So it'll need real work, and recently I was only able to contribute by doing changes that can be done quickly withouth needing to think a lot

**Comment by mikehardy** [comment]
pre-load != pre-render
I believe WebViews in the multi-process architecture may be limited to one per process? This is all medium- to far-future work, IMHO, vs loading a lighter font
Worth noting that loading a heavy font also leads to WebView crashes, so even if you get all that working @Arthur-Milchior with such heavy resource usage by (avoidable! there are lighter fonts) font usage, @ekumlin might just be set up to then crash, but quickly

**Comment by Arthur-Milchior** [comment]
To be honest, I don't care about font, I care about mathjax and media. So and there are still things that I'll pre-load. 
The other think I thought about was to simply add the elements in the current webview, by adding a js function into it to pre-render the content before it's needed
